[TB] Add reviews here (#152)
hassanobeid1994/tr_b_causal_2020: article/sections/_6_graph_discovery.tex

\section{Causal Discovery}
\label{sec:causal-discovery}

% How is causal discovery related to testing causal graphs?
Our previous section detailed a method for testing independence assumptions of one's causal graph.
As presented, this strategy requires one to first have a causal graph.
Indeed, this requirement is why Section \ref{sec:graph-construction} provides instructions on how to construct an initial causal graph using expert opinion.

> Our unstated presumption is that we will test our working causal graph.

Reviewer: Changed from working to proposed just in case some people understood working to be "valid" or correct, which could lead them to more confusion since you're talking about failing tests in the following sentence.
```suggestion
Our unstated presumption is that we will test our proposed causal graph.
```

Author: Done! See PR [here](https://github.com/hassan-obeid/tr_b_causal_2020/pull/155) as it'll be easier for me to merge.

> If any of our tests fail, we will then revise the graph until its assumptions appear plausible.

Reviewer: Changed to defensible because really anything could be plausible. I understand where you're coming from, meaning that given our data, our assumptions are plausible. I think defensible gives ownership of the assumptions to the researcher.
```suggestion
If any of our tests fail, we will then revise the graph until its assumptions appear defensible.
```

Author: Done! See PR [here](https://github.com/hassan-obeid/tr_b_causal_2020/pull/155) as it'll be easier for me to merge.

Essentially, we postulate, test, and edit our causal graph until the data conform to the graph's assumptions.
If this iterative discovery process sounds tedious, that is because it can be!

> In this section, we will discuss how we can avoid such repetitive, manual graph editing.

Reviewer: In my mind these two things are distinct yet complimentary tasks. You can't edit without testing. You mentioned it earlier, so this might be an overkill.

```suggestion
In this section, we will discuss how we can avoid such repetitive, manual graph editing and testing.
```

Author: Done! See PR [here](https://github.com/hassan-obeid/tr_b_causal_2020/pull/155) as it'll be easier for me to merge.

> Instead, we use our computers algorithmic capabilities.

Reviewer: I think this sentence should be removed. The following sentence is duplicate to me. Unless you think there are some other algorithmic capabilities that do this work.
```suggestion
```

Author: Done! See PR [here](https://github.com/hassan-obeid/tr_b_causal_2020/pull/155) as it'll be easier for me to merge.

Specifically, this section describes causal discovery:
algorithmically inferring our causal graph from data.
Here, we detail why causal discovery is important;
we provide a brief overview of the main concepts in causal discovery;
and we show the results of using causal discovery algorithms on the dataset described in Section \ref{sec:graph-importance}.

\subsection{Why use causal discovery?}
\label{sec:why-causal-discovery}

% Why is causal discovery important?
% Promotes robustness and understanding of results
% Inspires creation and editing of expert-opinion graph
% Aids characterization of posterior uncertainty
As a topic of study, causal discovery is important for numerous reasons.
Three of these reasons are the following.
First, causal discovery promotes robustness and understanding of our causal effect estimates.
For example, we can use causal discovery to understand how our prior beliefs affect our inferences.
To do so, we could compare our inferred causal effects under expert-opinion versus data-driven causal graphs.

Secondly, causal discovery helps inspire the creation and editing of expert-opinion graphs.

> Indeed, it's typically easier to edit than to create from scratch.

Reviewer: ```suggestion
Indeed, it is typically easier to edit than to create from scratch.
```

Author: Done! See PR [here](https://github.com/hassan-obeid/tr_b_causal_2020/pull/155) as it'll be easier for me to merge.

> Graphs found via causal discovery can spark ideas for one's own graph if they feature different causal relations than are present in one's own graph, and they provide a starting point for graph revision.

Reviewer: What I understand from this paragraph is that causal discovery helps in both the creation of initial graphs. But this sentence refers back to some initial causal graph that was created by the researcher. Was the intent to refer to both things?

I don't know why I am struggling to separate editing a causal graph from checking the robustness of results.

Author: Let's talk more about this on Sunday.

Author: Actually, does the current edit work for you? I changed to:

```suggestion
If we have not already created our own causal graphs, then we will find that it's typically easier to edit discovered graphs than to create from scratch.
Alternatively, graphs found via causal discovery can spark revisions if we have already created our own graphs.
This is especially likely when the discovered graphs feature different causal relations than are present in our own graph.
```

The diff is messed up (missing original line 36 in red) but hopefully you get the point.

%Randomize-then-optimize.
Alternatively, we could use yet another randomize-then-optimize \citep{bardsley_2014_randomize, orabona_2014_measure} approach to sampling from an approximate posterior distribution.
Here, one would sample from a prior on entries of the matrix representation of a causal graph.
Semantically, we sample constraints such as ``$X \rightarrow Y$ (MUST | MUST NOT) be in the causal graph''.
We ``sample'' from the approximate posterior by running the causal discovery algorithm on the original dataset, with the inclusion of these randomly generated constraints.
And, of course, we can consider hybrids of these two posterior approximation schemes.

With these posterior approximation methods, we can generate causal graphs for all the purposes mentioned above:
\begin{itemize}
   \item for criticism and inspiration,
   \item for distributional analyses of one's causal effect estimates conditional on each sampled graph, and
   \item for distributional analyses of the causal graphs themselves.
   \newline
   I.e., how certain are we of any one causal graph?

> \end{itemize}

Reviewer: Should we reorder these to be in the same order as we present them above?

Author: Done! See PR [here](https://github.com/hassan-obeid/tr_b_causal_2020/pull/155) as it'll be easier for me to merge.

\subsection{Overview of causal discovery algorithms}
\label{sec:discovery-overview}

% Pointers to thorough review papers

> The following subsection presents a non-exhaustive overview of causal discovery algorithms.

Reviewer: ```suggestion
This subsection presents a non-exhaustive overview of causal discovery algorithms.
```

Author: Done! See PR [here](https://github.com/hassan-obeid/tr_b_causal_2020/pull/155) as it'll be easier for me to merge.

Crucially, we rely heavily on review papers such as \citet{glymour_2019_review} and \citet{spirtes_2016_causal} to fill in our gaps.

% Types of causal discovery algorithms
Overall, there are three classes of causal discovery algorithms.
One class attempts to directly infer the marginal and conditional independences in one's causal system.
That is, this class of algorithms identifies a so-called Markov Equivalence Class (MEC) of graphs.
Considering the discussion of independence testing in Section \ref{sec:graph-testing}, this class of algorithms is perhaps best understood as repeated and systematic independence tests.
These causal discovery algorithms are constraint-based algorithms, because the observed independencies represent constraints that define the space of plausible causal graphs for the dataset.
Common constraint-based algorithms include the Peter-Clarke (PC) algorithm and the Fast Causal Inference (FCI) algorithm \citep{glymour_2001_causation}.
The PC algorithm assumes no unobserved confounding variables, whereas the FCI allows for (and sometimes infers) the presence of such unobserved confounders.

> % DESCRIBE SCORE BASED DISCOVERY ALGORITHMS

Reviewer: What are these scores? Are there example we can give of the scores? Just names of score metrics, or what the scores quantify?

Author: Usually just BIC. See PR [here](https://github.com/hassan-obeid/tr_b_causal_2020/pull/155) as it'll be easier for me to merge.

Author: And I added the specification of scores in the paper, not just on github comments lol.

Nonetheless, combinations of score and constraint-based ideas have outperformed methods from either class alone \citep{glymour_2019_review}.

% DESCRIBE FCM ALGORITHMS
Lastly, the third class of algorithms estimates Functional Causal Models (FCM) \citep{goudet_2018_learning} for each of the variables in our system.
The defining characteristic in such algorithms is that they exploit asymmetries in the residuals of models for the hypothesized relationships of $X \rightarrow Y$, $Y \rightarrow X$.
In particular, the residuals will be independent of the hypothesized cause in only one of the two potential models.
As noted in the description of such algorithms, they are especially useful for discovering causal relationships between pairs of variables.
In other words, discovery methods based on FCMs enable orientation of a graph's undirected or bi-directed edges.
By orienting these edges, FCMs reduce ambiguity over whether $X \rightarrow Y$ or $Y \rightarrow X$.
This orienting capability can be usefully combined with the previous causal discovery algorithms that infer classes graphs with equivalent marginal and conditional independencies.

> And of course, extensions of these techniques exist for inference in the presence of unobserved confounding \citep[Sec. 6]{goudet_2018_learning} and for learning an entire causal graph as opposed to dealing solely with variable pairs \citep{zheng_2020_learning}.

Reviewer: No comments here, but sheesh...people do work!

Author: Right? The amount of work done that we've never even scratched the surface in terms of using/investigating is massive!

\subsection{An application of causal discovery}
\label{sec:discovery-application}
To demonstrate the methods in this section, we chose to use the simplest causal discovery algorithm: the PC algorithm \citep{glymour_2001_causation}.

> For practitioners who may reasonably start with the simplest method they can find and increase methodological complexity as needed, this should be an illuminating starting point.

Reviewer: This sentence sounds funky to me so edited it to the following
```suggestion
For practitioners who may reasonably start with the simplest method available and increase methodological complexity as needed, this should be an illuminating starting point.
```

Author: Done! See PR [here](https://github.com/hassan-obeid/tr_b_causal_2020/pull/155) as it'll be easier for me to merge.

> Moreover, as noted above, the PC algorithm assumes that one has no unobserved confounders.
> As noted in Section \ref{sec:testing-demonstration}, we suspect that our causal graph for the drive alone utility of our dataset contains unobserved confounding.
> Accordingly, using the PC algorithm on our example allows us to observe how the algorithm behaves under violation of its assumptions.

Reviewer: ```suggestion
As noted in Section \ref{sec:testing-demonstration}, we suspect that our causal graph for the drive alone utility of our dataset contains unobserved confounding.
Given the PC algorithm assumes that one has no unobserved confounders, using it on our example allows us to observe how the algorithm behaves under violation of its assumptions.
```

Author: Done! See PR [here](https://github.com/hassan-obeid/tr_b_causal_2020/pull/155) as it'll be easier for me to merge.

Does the algorithm fail gracefully and point towards violations of its assumptions, or does it return incorrect results with certainty?

To begin, it helps to understand the general procedure of the PC algorithm. The algorithm uses all our observed variables to infer a skeleton of a causal graph.
That is, the algorithm attempts to infer an undirected graph that denotes which variables relate to which other variables, ignoring the directionality of causation between them.
Then, after inferring a skeleton, the algorithm attempts to orient the undirected edges as much as possible.
Optimistically, we end up with a fully directed acyclic graph.
In other cases, we recover a mix of directed and undirected edges, denoting a MEC of graphs with the same independence properties.

> In the final case, we recover an undirected causal graph.

Reviewer: ```suggestion
In the worst case, we recover an undirected causal graph.
```

Reviewer: hmmm thinking about how a reader could interpret worst here and second doubting my suggestion. Is it really bad that we recover an undirected causal graph? I mean sure it's bad for us because it means more work, but is it objectively or scientifically bad?

Author: I don't think it's necessarily scientifically bad. It means one has to use other techniques to calculate causal effects. I think I'll stick with final to avoid any incorrect connotations.
